src/kOS.Safe/Compilation/KS/Compiler.cs

```diff
@@ -413,76 +413,85 @@ private void PreProcessChildNodes(ParseNode node)
         private void PreProcessOnStatement(ParseNode node)
         {
             NodeStartHousekeeping(node);
+            nextBraceIsFunction = true; // triggers aren't really functions but act like it a lot.
+
             int expressionHash = ConcatenateNodes(node).GetHashCode();
             string triggerIdentifier = "on-" + expressionHash.ToString();
             Trigger triggerObject = context.Triggers.GetTrigger(triggerIdentifier);
-            triggerObject.SetTriggerVariable(GetIdentifierText(node));
 
             currentCodeSection = triggerObject.Code;
-            AddOpcode(new OpcodePush(triggerObject.VariableNameOldValue));
-            AddOpcode(new OpcodePush(triggerObject.VariableName));
+            // Put the old value on top of the stack for equals comparison later:
+            AddOpcode(new OpcodePush(triggerObject.OldValueIdentifier));
+            AddOpcode(new OpcodeEval());
+            // eval the expression for the new value, and leave it on the stack twice.
+            VisitNode(node.Nodes[1]);
+            AddOpcode(new OpcodeEval());
+            AddOpcode(new OpcodeDup());
+            // Put one of those two copies of the new value into the old value identifier for next time:
+            AddOpcode(new OpcodePush(triggerObject.OldValueIdentifier));
+            AddOpcode(new OpcodeSwap());
+            AddOpcode(new OpcodeStoreGlobal());
+            // Use the other dup'ed copy of the new value to actually do the equals
+            // comparison with the old value that's still under it on the stack:
             AddOpcode(new OpcodeCompareEqual());
-            AddOpcode(new OpcodeLogicNot());
-            Opcode branchOpcode = AddOpcode(new OpcodeBranchIfFalse());
+            OpcodeBranchIfFalse branchToBody = new OpcodeBranchIfFalse();
+            branchToBody.Distance = 3;
+            AddOpcode(branchToBody);
+            AddOpcode(new OpcodePush(true));       // wasn't triggered yet, so preserve.
+            AddOpcode(new OpcodeReturn((short)0)); // premature return because it wasn't triggered
 
             // make flag that remembers whether to remove trigger:
             // defaults to true = removal should happen.
-            string triggerRemoveVarName = "$remove-" + triggerIdentifier;
-            PushTriggerRemoveName(triggerRemoveVarName);
-            AddOpcode(new OpcodePush(triggerRemoveVarName));
-            AddOpcode(new OpcodePush(true));
+            string triggerKeepName = "$keep-" + triggerIdentifier;
+            PushTriggerKeepName(triggerKeepName);
+            AddOpcode(new OpcodePush(triggerKeepName));
+            AddOpcode(new OpcodePush(false));
```

Comment:
The logic was inverted from before.  Before it was:

   keep trigger active until the end of the body, then remove trigger unless preserved.

The new logic is:

    remove trigger right away at the top, then re-add it at the bottom by choosing to return true or false from the trigger depending on if it wants to be preserved.

Now it's the Return opcode itself that chooses to re-insert the trigger, rather than it happening in the trigger's code.  This is to force it to be atomic, else the IPU boundary could have been hit in-between the re-insertion and the return.

(This same comment applies to the WHEN statement logic further down.)